### PR TITLE
Moses/4 live reload

### LINKED ISSUE #4: Live Reloading
Having to restart the app each time you make a change sucks. Live reloading would be a great feature to have.

If I were to suggest some things, I'd first suggest leveraging `uvicorn`'s built-in `--reload` parameter which will automatically reload your app when something changes.

You can inject the javascript into the page's head with the `html_head` argument in `QuickHTML`
```py
app = QuickHTML(html_head=Script(...))
```

Though we'll probably want another argument that can do that for us instead

Tasks:
- [ ] An argument added to the `QuickHTML` class to allow for live reloading
- [ ] A JS snippet to be injected into the rendered HTML page that will refresh the page when the codebase changes
  - Ideally this uses a websocket (which we haven't actually implemented support yet for)

### PR DESCRIPTION
**Pull request by moseye**
Closes #4 

Now with tests!

### CODE REVIEW

File: src/rapidhtml/app.py [hunk 1]
```diff
@@ -1,13 +1,40 @@
...
+sock = new WebSocket(`ws://${window.location.host}/live-reload`);
+sock.onopen = function (event) {
+    console.log(`connected`);
+    active = true;
+};
+
+sock.onclose = function (event) {
+    console.log(`disconnected... reloading.`);
+    if (active) {
+        setTimeout(function () {
```
Comment (ThomasJRyan) [inline src/rapidhtml/app.py:24]: Is `setTimeout` necessary? I can remove it with seemingly no ill effects. If it is necessary, we need to reduce the timeout to something much smaller (I was able to set it to 0 with no issues). Having to wait five seconds for the page to reload isn't ideal

File: src/rapidhtml/app.py [hunk 2]
```diff
@@ -1,13 +1,40 @@
...
 
 from rapidhtml.tags import Script
-from rapidhtml.routing import RapidHTMLRouter
+from rapidhtml.routing import RapidHTMLRouter, RapidHTMLWSEndpoint
+
+JS_RELOAD_SCRIPT = """           
+let active = false;
+sock = new WebSocket(`ws://${window.location.host}/live-reload`);
+sock.onopen = function (event) {
+    console.log(`connected`);
```
Comment (ThomasJRyan) [inline src/rapidhtml/app.py:17]: Might as well be cute about it
```suggestion
    console.log(`Connected to the RapidHTML development server!`);
```

File: src/rapidhtml/app.py [hunk 3]
```diff
@@ -1,13 +1,40 @@
...
+    active = true;
+};
+
+sock.onclose = function (event) {
+    console.log(`disconnected... reloading.`);
+    if (active) {
```
Comment (ThomasJRyan) [inline src/rapidhtml/app.py:23]: Might not need `active` either actually? Same thing, no issues for me locally

File: src/rapidhtml/app.py [hunk 4]
```diff
@@ -20,17 +47,26 @@ class RapidHTML(Starlette):
         - HTMX
     """
 
-    def __init__(self, *args, html_head: typing.Iterable = None, **kwargs) -> None:
+    def __init__(
+        self, *args, html_head: typing.Iterable = None, reload: bool = False, **kwargs
+    ) -> None:
```
Comment (ThomasJRyan) [inline src/rapidhtml/app.py:52]: Probably time for a docstring
```suggestion
    ) -> None:
    """Initializes the RapidHTML application.

        Args:
            html_head (typing.Iterable, optional): Tags to inject into each 
                page's <head>. Defaults to None.
            reload (bool, optional): Enables live-reloading. Defaults to False.
    """
```

**Review by ThomasJRyan — CHANGES_REQUESTED** [review]
Is there any way for us to unittest this...

File: src/rapidhtml/app.py [hunk 5]
```diff
@@ -20,17 +47,26 @@ class RapidHTML(Starlette):
...
+        if reload:
+            self.html_head += (Script(JS_RELOAD_SCRIPT),)
+            self.router = RapidHTMLRouter(html_head=self.html_head)
+            self.router.add_websocket_route("/live-reload", _ReloadSocket)
+        else:
+            self.router = RapidHTMLRouter(html_head=self.html_head)
 
     def serve(self, *args, **kwargs):
-        uvicorn.run(self, *args, **kwargs)
+        appname = Path(inspect.currentframe().f_back.f_globals.get("__file__", "")).stem
```
Comment (rashidnhm) [inline src/rapidhtml/app.py:68]: Would:
`Path(__file__).stem` work? Or do you HAVE TO go one frame up?

Comment (rashidnhm) [reply src/rapidhtml/app.py:68]: Tried it, won't work

**Review by ThomasJRyan — DISMISSED** [review]
Would also be nice if something was written about this in the docs

**Comment by ThomasJRyan** [comment]
This requires `app.serve()` to be inside of `if __name__ == "__main__"`. Would be nice if that could be worked around somehow

File: src/rapidhtml/app.py [hunk 6]
```diff
@@ -20,17 +47,26 @@ class RapidHTML(Starlette):
...
-        self.router = RapidHTMLRouter(html_head=self.html_head)
+
+        if reload:
+            self.html_head += (Script(JS_RELOAD_SCRIPT),)
+            self.router = RapidHTMLRouter(html_head=self.html_head)
+            self.router.add_websocket_route("/live-reload", _ReloadSocket)
+        else:
+            self.router = RapidHTMLRouter(html_head=self.html_head)
 
     def serve(self, *args, **kwargs):
```
Comment (rashidnhm) [inline src/rapidhtml/app.py:67]: It's possible for the user to pass `reload=` here as well. We should pop it out and raise a Warning.

**Comment by moseye** [comment]
> This requires `app.serve()` to be inside of `if __name__ == "__main__"`. Would be nice if that could be worked around somehow

Unfortunately not possible due to how uvicorn is handling reloading the app. From my understanding the current flow is as such
1. We start main.py, which calls app.serve()
2. Uvicorn runs, and sees live-reloading is selected, configures sockets, and then goes to reload and attempt to load main.py again.
3. If main.py has an app.serve, it starts this process all over again, and since the configured IP address and socket are already used, it errors out.

I tried to fix this but it was becoming very hacky and jank. If either of you @rashidnhm @ThomasJRyan have any ideas im open, but I think enforcing an if main block is not a big deal.

File: src/rapidhtml/app.py [hunk 7]
```diff
@@ -20,17 +38,43 @@ class RapidHTML(Starlette):
...
-    def serve(self, *args, **kwargs):
-        uvicorn.run(self, *args, **kwargs)
+        if reload:
+            self.html_head += (Script(JS_RELOAD_SCRIPT),)
+            self.router = RapidHTMLRouter(html_head=self.html_head)
+            self.router.add_websocket_route("/live-reload", _ReloadSocket)
+        else:
+            self.router = RapidHTMLRouter(html_head=self.html_head)
+
+    def serve(self, appname=None, host="127.0.0.1", port=8001, *args, **kwargs):
```
Comment (ThomasJRyan) [inline src/rapidhtml/app.py:67]: I think it should be 8000. It's the default for uvicorn and I initially used 8001 due to the port being taken and messed up by pushing the wrong example code
```suggestion
    def serve(self, appname=None, host="127.0.0.1", port=8000, *args, **kwargs):
```

File: tests/test_live_reload.py
Comment (ThomasJRyan) [inline tests/test_live_reload.py:1]: Would the TestClient be applicable here?
https://www.starlette.io/testclient/

If not it's fine. So long as the tests cover what they're supposed to all's good

**Review by ThomasJRyan — DISMISSED** [review]

File: tests/test_live_reload.py
Comment (moseye) [reply tests/test_live_reload.py:1]: Tried originally, seems unable to use live-reloading unfortunately.

**Review by ThomasJRyan — APPROVED** [review]

**Pull request merged**